elide-async/src/main/java/com/yahoo/elide/async/service/AsyncQueryThread.java

```diff
@@ -99,16 +135,98 @@ else if (queryObj.getQueryType().equals(QueryType.GRAPHQL_V1_0)) {
+                throw new IllegalStateException("Unable to store async results for download");
+            }
+            queryResultObj = resultStorageEngine.storeResults(queryResultObj, tempResult, queryObj.getId());
+            queryResultObj.setResponseBody(resultStorageEngine.generateDownloadUrl(requestURL,
+                    queryObj.getId()).toString());
+        }
+
         return queryResultObj;
     }
 
     /**
-     * This method parses the url and gets the query params.
+     * This method calculates the number of records from the response with JSON API.
+     * @param jsonStr is the response.getBody() we get from the response
+     * @return rec is the recordCount
+     */
+    protected Integer calculateRecordsJSON(String jsonStr) {
+        return JsonPath.read(jsonStr, "$.data.length()");
+    }
+
+    /**
+     * This method calculates the number of records from the response with GRAPHQL API.
+     * @param jsonStr is the response.getBody() we get from the response
+     * @return rec is the recordCount
+     * @throws IOException Exception thrown by JsonPath
+     */
+    protected Integer calculateRecordsGraphQL(String jsonStr) throws IOException {
+        List<Integer> countList = JsonPath.read(jsonStr, "$..edges.length()");
+        Integer count = countList.size() > 0 ? countList.get(0) : 0;
+        return count;
+    }
+
+    /**
+     * This method checks if the json string is error message.
+     * @param jsonStr is the response.getBody() we get from the response
+     * @return is the message an error message
+     */
+    protected boolean checkJsonStrErrorMessage(String jsonStr) {
+        boolean isError = false;
+        try {
+            isError = (Integer) JsonPath.read(jsonStr, "$.errors.length()") >= 1;
+        } catch (PathNotFoundException e) {
+            //ignore when not an error message
+            log.debug(e.getMessage());
+        }
+        return isError;
+    }
+
+    /**
+     * This method converts the JSON response to a CSV response type.
+     * @param jsonStr is the response.getBody() of the query
+     * @return retuns a string which nis in CSV format
+     * @throws IllegalStateException Exception thrown
+     */
+    protected String convertJsonToCSV(String jsonStr) {
+        if (jsonStr == null) {
+            return null;
+        }
+        StringBuilder str = new StringBuilder();
+        JFlat flatMe = new JFlat(jsonStr);
+        List<Object[]> json2csv;
+        try {
+            json2csv = flatMe.json2Sheet().headerSeparator("_").getJsonAsSheet();
+            for (Object[] obj : json2csv) {
+                String objString = Arrays.toString(obj);
+                if (objString != null) {
+                    objString = objString.substring(1, objString.length() - 1);
+                }
+                str.append(objString);
+                str.append(System.getProperty("line.separator"));
+            }
+        } catch (Exception e) {
```

Comment:
JFlat throws the checked exception (Exception), so I am catching it to convert to IllegalException instead of letting Exception being thrown to the calling function, and then it is thrown back up the chain.